README.md

```diff
@@ -157,136 +139,121 @@ symbols, it will fail at class definition time. This program will throw:
 
-`Protocol.implement` accepts zero or more protocols following the class.
+Having `implFor` allows extending existing prototypes for one's own purposes
+while preventing multiple conflicting implementations using the same protocol
+symbols. With this format, there's only two ways to implement a protocol:
 
-```js
-protocol I {}
-protocol K {}
+1. Using the new `implements` keyword in `class` bodies, which allows new class definitions to use a pre-existing protocol.
+2. Using `implFor` to make an existing class comply with the protocol.
 
-// these are all the same
-let C = Protocol.implement(Protocol.implement(class C {}, I), K);
-let C = Protocol.implement(class C {}, I, K);
-class C implements I, K {}
-```
+If you do not own one or the other, you are unable to implement the protocol. It
+is impossible to define individual protocol properties and methods in class
+definitions without using the `implements` keyword. This guarantees coherence,
+which means a protocol can only be implemented once and only once, globally.
 
 ### `implements` operator
 
-The `implements` operator returns `true` if and only if a given class has all
-of the symbols required to implement a given protocol as well as the methods
-provided by implementing the protocol.
+The `implements` operator returns `true` iff a given class has implemented a
+protocol, or if a protocol has provided an `implFor` for an existing class.
+There is no way to define individual protocol methods on a class without using
+one of these two methods.
 
 ```js
 protocol I { a; b() {} }
 protocol K { a; b() {} }
 
-class C {
-  [I.a]() {}
-}
-C implements I; // false
-C implements K; // false
-
-class D implements I {
-  [I.a]() {}
+class D {
+  implements I {
+    a() {}
+  }
 }
 D implements I; // true
 D implements K; // false
 
-class E {
-  [I.a]() {}
-  [I.b]() {}
+class F {
+  implements I {
+    a() {}
+  }
+  implements K {
+    a() {}
+  }
 }
-E implements I; // true
-E implements K; // false
+D implements I; // true
+D implements K; // true
 ```
 
 ### static methods
 
-Some protocols would like to provide methods for the constructor instead of the
-prototype. Use the `static` modifier for this.
+Static protocol methods are added to the protocol object itself, not to classes
+that implement it:
 
 ```js
 protocol A {
-  static b() {}
+  static b = 'b'
+  static c() { return 'c'; }
 }
 
-class C implements A { }
-C[A.b]();
+A.b() // 'b'
+A.c() // 'c'
 ```
 
-Similarly, require a protocol symbol to be on the constructor instead of the
-prototype using the `static` modifier.
+### Legacy string-based protocols
 
-```js
-protocol Monoid {
-  concat;
-  static identity;
-}
-```
-
-### required strings
-
-In order to allow representation of "legacy" protocols which depend on strings,
-the protocol definition syntax allows for a string literal to be used in place of
-the identifier which declares a required symbol.
+There is no way to implement or enforce classic property-based protocols using
```

Comment:
This imo is a blocker, which is why it was added to this proposal in the first place. There needs to be a way to represent existing Thenable, Arraylike etc, without having to install new symbols and exhaustively list everything that implements it.

Specifically, i need to be able to have a protocol that represents everything `await` will consume.

Reply:
I wanted to make this an exercise in whether this was actually necessary or sensible for this construct, or just something that had snuck in because it seemed to make sense kinda.

It's important to note that coherence guarantees go right out the window if this is allowed, and I'm wondering if it would be best to restrict the scope of this feature a bit.

Reply:
I don’t find coherence guarantees to be valuable (at least, valuable enough to trump other concerns), and i don’t think a facility for protocols should be added to the language that can’t support all of the existing protocols - thenables, iterables, arraylikes, toStringables, etc.

Reply:
This feedback is not particularly actionable. Can you analyze your needs, keep an open mind, and propose potential alternatives without dismissing existing goals right off the bat? That would be way more useful to me as far as feedback goes.

And frankly, opening any conversation in this committee with threats to block based on a WIP, ongoing discussion mostly fills me with stress and distrust and doesn't particularly make me want to engage further.

Reply:
First, I apologize that my wording has filled you with those feelings; that wasn't my intent, but my intent isn't important.

Earlier you said you wanted to have an exercise "in whether this was actually necessary or sensible for this construct, or just something that had snuck in because it seemed to make sense kinda." - per #19, this is a "heavily requested feature from some delegates", myself included, which imo makes it necessary, so I feel like this "exercise" is indeed dismissing existing goals right off the bat for the sake of a goal that isn't present in the current proposal.

The language already *has* implicit protocols: thenable (anything with a function under the string "then"); iterable (anything with a function under Symbol.iterable); arraylike (anything with an integer-coercible `length` property); plus toString, valueOf, toJSON, and a few others. It doesn't make any sense to me to propose a feature for explicit protocols (despite that I want them very badly) if it doesn't mesh with what's already in the language. I don't have a particular implementation in mind, and I'm very open to whatever can be proposed to meet that criteria.

Reply:
> dismissing existing goals right off the bat for the sake of a goal that isn't present in the current proposal

Sure, but other delegates (@wycats, @dherman) expressed that ensuring/promoting coherence was also a critical feature for this proposal. It at least needs to be explored and addressed as best as possible, which includes exploring it to the lengths that it also drops other important features. The end result will hopefully be a proposal which does the best job addressing as many concerns as possible.

As for representing existing implicit protocols: keep in mind that this is a very difficult requirement. The existing protocols were all added in an ad-hoc manner at different times with different designs. We will do our best to provide the ability to represent the most structured ones, but things like array-like would likely complicate this proposal beyond what is reasonable with little general benefit.

Reply:
@michaelficarra thanks, that’s useful to know (is it documented elsewhere in the repo?)

Reply:
Not in this repo, sorry. From the [Sept 2017 meeting notes](https://github.com/tc39/tc39-notes/blob/master/es8/2017-09/sep-26.md#11iva-first-class-protocols) where this was presented for stage 1:

> DH: A few reactions. I'm not sure I share aversion to inheritance - this is about fitting in with object mechanisms we have. The cowpath of existing mixin pattern is an important signal. Primary concerns though: the convenience of thing.map over thing[SomeProtocol.map].
> 
> MF: Were the examples in slides not convenient enough?
> 
> DH: It's hard to beat thing.map. My other concern is wanting to allow separate pieces of code to allow implementations of interfaces independently. Show us an example protocol, the ways in which instance coherence blows up and what this does to resolve that. How do you solve the "instance coherence" problem (which has been extensively studied in the industry)?
> 
> MF: Thanks that's great feedback.
> 
> YK: I'm worried that multiple pieces of code implement the exact same protocol. The "who got there first" problem isn't solved for laws about types.
> 
> MF: Yes, you can implement the same laws but with different performance characteristics.
> 
> YK: Yes, so in Rust you either implement your protocol for a foreign type or a foreign protocol for your type. This isnt the case here?
> 
> DH: This is out there - but it has come up before that we need way to talk about "packages" in general. When people write JavaScript there is a set of implicit semantics that strongly defines a "package". For example Rust brings in the notion of crates as a first class piece of the language.

Unfortunately, not all of the nuance of the discussion was reflected in the notes, but you get the gist. We initially thought defining some sort of package boundary was the only way to solve this.